packages/fxa-content-server/app/scripts/lib/email-domain-validator.js

```diff
@@ -0,0 +1,116 @@
+/* This Source Code Form is subject to the terms of the Mozilla Public
+ * License, v. 2.0. If a copy of the MPL was not distributed with this
+ * file, You can obtain one at http://mozilla.org/MPL/2.0/. */
+
+/**
+ * This module valdiates an email address's domain through DNS lookups as well
+ * as display an error/tooltip for the email address field.
+ *
+ * The validation is performed after the existing account email check during
+ * the form submission process. At this point, the email address has gone
+ * through: 1) regex validation, 2) Mailcheck typo suggestion (which the user
+ * can ignore), and 3) existing email address check.
+ *
+ * There are three possible validation results: 'MX', 'A', and 'none'.
+ *  - 'MX': MX record exists, proceed
+ *  - 'A': no MX, but A record exists, display tooltip but allow submission
+ *  - 'none: neither MX or A records found, block submission
+ *
+ * If the validation request itself fails, we allow the submission to
+ * continue.
+ */
+
+import AuthErrors from './auth-errors';
+import Tooltip from '../views/tooltip';
+import TopEmailDomains from './top-email-domains';
+import xhr from './xhr';
+const t = msg => msg;
+
+const EMAIL_VALIDATION_ENDPOINT = '/validate-email-domain';
+// This allows us to distinguish between our own rejection and a server error in
+// a `catch` that'll catch both.
+const REJECTION_CODE = '__INVALID_EMAIL_DOMAIN__';
+const MISTYPED_EMAIL_MESSAGE = t('Mistyped email?');
+
+const resolveDomain = domain =>
+  xhr.ajax({
+    contentType: 'application/json',
+    dataType: 'json',
+    data: { domain },
+    type: 'POST',
+    url: EMAIL_VALIDATION_ENDPOINT,
+  });
+
+const checkEmailDomain = ($element, view) => {
+  const showInvalidDomainError = domain => {
+    const invalidDomainError = AuthErrors.toInvalidEmailDomainError(domain);
+    view.showValidationError($element, invalidDomainError);
+    view.logEvent('emailDomainValidation.fail');
+  };
+
+  const showTooltip = () => {
+    const tooltip = new Tooltip({
+      dismissible: true,
+      extraClassNames: 'tooltip-suggest tooltip-error',
+      invalidEl: $element,
+      type: 'emailDomainValidation',
+      unsafeMessage: MISTYPED_EMAIL_MESSAGE,
+    });
+
+    tooltip.on('destroyed', () =>
+      view.logEvent('emailDomainValidation.dimissed')
+    );
+    tooltip.render();
+    view.logEvent('emailDomainValidation.fallback');
+  };
+
+  return new Promise((resolve, reject) => {
+    const emailAddress = $element.val();
+    const [, domain] = emailAddress.split('@');
+
+    if (TopEmailDomains.has(domain)) {
```

Comment:
Can you leave a comment here about if the domain is part of a known good list, then skip the network check?